sopel/irc/isupport.py

```diff
@@ -0,0 +1,384 @@
+
+    return value
+
+
+def _map_items(parser=str, map_separator=',', item_separator=':'):
+    @functools.wraps(parser)
+    def wrapped(value):
+        items = sorted(
+            item.split(item_separator)
+            for item in value.split(map_separator))
+
+        return tuple(
+            (k, parser(v) if v else None)
+            for k, v in items
+        )
+    return wrapped
+
+
+def _parse_chanmodes(value):
+    items = value.split(',')
+
+    if len(items) < 4:
+        raise ValueError('Not enough channel types to unpack from %r.' % value)
+
+    # add extra channel mode types to their own tuple
+    # result in (A, B, C, D, (E, F, G, H, ..., Z))
+    # where A, B, C, D = result[:4]
+    # and extras = result[4]
+    return tuple(items[:4]) + (tuple(items[4:]),)
+
+
+def _parse_elist(value):
+    # letters are case-insensitives
+    return tuple(sorted(set(letter.upper() for letter in value)))
+
+
+def _parse_extban(value):
+    args = value.split(',')
+
+    if len(args) < 2:
+        raise ValueError('Invalid value for EXTBAN: %r.' % value)
+
+    prefix = args[0] or None
+    items = tuple(sorted(set(args[1])))
+
+    return (prefix, items)
+
+
+def _parse_prefix(value):
+    result = re.match(r'\((?P<modes>\S+)\)(?P<prefixes>\S+)', value)
+
+    if not result:
+        raise ValueError('Invalid value for PREFIX: %r' % value)
+
+    modes = result.group('modes')
+    prefixes = result.group('prefixes')
+
+    if len(modes) != len(prefixes):
+        raise ValueError('Mode list does not match for PREFIX: %r' % value)
+
+    return tuple(sorted(zip(modes, prefixes)))
+
+
+ISUPPORT_PARSERS = {
+    'AWAYLEN': int,
+    'CASEMAPPING': str,
+    'CHANLIMIT': _map_items(int),
+    'CHANMODES': _parse_chanmodes,
+    'CHANNELLEN': int,
+    'CHANTYPES': _optional(tuple),
+    'ELIST': _parse_elist,
+    'EXCEPTS': _optional(_single_character, default='e'),
+    'EXTBAN': _parse_extban,
+    'HOSTLEN': int,
+    'INVEX': _optional(_single_character, default='I'),
+    'KICKLEN': int,
+    'MAXLIST': _map_items(int),
+    'MAXTARGETS': _optional(int),
+    'MODES': _optional(int),
+    'NETWORK': str,
+    'NICKLEN': int,
+    'PREFIX': _optional(_parse_prefix),
+    'SAFELIST': _no_value,
+    'SILENCE': _optional(int),
+    'STATUSMSG': _optional(tuple),
+    'TARGMAX': _optional(_map_items(int)),
+    'TOPICLEN': int,
+    'USERLEN': int,
+}
+
+
+def parse_parameter(arg):
+    items = arg.split('=', 1)
+    if len(items) == 2:
+        key, value = items
+    else:
+        key, value = items[0], None
+
+    if key.startswith('-'):
+        # ignore value for removed parameters
+        return (key, None)
+
+    parser = ISUPPORT_PARSERS.get(key, _optional(str))
+    return (key, parser(value))
+
+
+class ISupport(object):
+    """Storage class for IRC's ``ISUPPORT`` feature.
+
+    An instance of ``ISupport`` can be used as a read-only dict, to store
+    feature advertised by the IRC server::
```

Comment:
```suggestion
    features advertised by the IRC server::
```

I missed this before! Caught while drafting changelog entry and building the docs to make sure my cross-links would be correct. 😁